Implemented reverse mode for inverse

IvanYashchuk opened the pull request:
## Summary

This PR implements reverse mode differentiation for `stan::math::inverse`.

## Tests

Added tests for 0x0, 1x1 input; input containing NaN, Inf; singular matrix input. Copied and modified test_ad_unary from test_ad_test.cpp

## Side Effects

Are there any side effects that we should be aware of?
No.

## Checklist

- [x] Math issue #989 

- [x] Copyright holder: Ivan Yashchuk

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

IvanYashchuk commented on stan/math/rev/mat/fun/inverse.hpp:
```diff
@@ -0,0 +1,87 @@
+#include <stan/math/prim/mat/fun/typedefs.hpp>
+#include <stan/math/rev/core.hpp>
+#include <stan/math/rev/mat/fun/typedefs.hpp>
+#include <stan/math/prim/mat/err/check_square.hpp>
```
Done.

bob-carpenter commented:
Did this use the new testing framework?  

It just got merged and I used inverse as an example in the tests of the tests.

It's in math in test/unit/math/test_ad.hpp and test_ad_test.cpp for the tests of the tester.

> On Sep 3, 2019, at 11:23 AM, Stan buildbot <notifications@github.com> wrote:
> 
> (stat_comp_benchmarks/benchmarks/gp_pois_regr/gp_pois_regr.stan, 0.99)
> (stat_comp_benchmarks/benchmarks/low_dim_corr_gauss/low_dim_corr_gauss.stan, 0.99)
> (stat_comp_benchmarks/benchmarks/irt_2pl/irt_2pl.stan, 1.0)
> (stat_comp_benchmarks/benchmarks/pkpd/one_comp_mm_elim_abs.stan, 1.01)
> (stat_comp_benchmarks/benchmarks/eight_schools/eight_schools.stan, 0.99)
> (stat_comp_benchmarks/benchmarks/gp_regr/gp_regr.stan, 1.01)
> (compilation, 1.0)
> (stat_comp_benchmarks/benchmarks/arK/arK.stan, 1.01)
> (stat_comp_benchmarks/benchmarks/low_dim_gauss_mix_collapse/low_dim_gauss_mix_collapse.stan, 1.0)
> (stat_comp_benchmarks/benchmarks/low_dim_gauss_mix/low_dim_gauss_mix.stan, 1.01)
> (stat_comp_benchmarks/benchmarks/sir/sir.stan, 1.0)
> (stat_comp_benchmarks/benchmarks/pkpd/sim_one_comp_mm_elim_abs.stan, 1.0)
> (stat_comp_benchmarks/benchmarks/garch/garch.stan, 1.03)
> (stat_comp_benchmarks/benchmarks/gp_regr/gen_gp_data.stan, 0.99)
> (stat_comp_benchmarks/benchmarks/arma/arma.stan, 1.01)
> Result: 1.00299271903
> Commit hash: dadcc07
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

rok-cesnovar commented:
The tests for rev/mat/funinverse are already in place: https://github.com/stan-dev/math/blob/develop/test/unit/math/rev/mat/fun/inverse_test.cpp

They dont use your new testing framework yet.

bob-carpenter commented:
I'm not reviewing, but I wouldn't accept this as is.  The tests
are missing just about everything:

* behavior on NaN and infinite components

* behavior on 0 x 0 and 1 x 1 matrices

* exception behavior on non-square matrices

* any other exception behavior on inputs (don't know if there is anything else)

* behavior of derivatives (as is, being tested in the x[0, 1] slot for a 2x2 matrix
  and in testing gradients are all 1 in sum(inverse(inverse(x)) (because it's
  equivalent to sum(x)---would've saved me some time if that was documented)

I know they were missing before the refactor, but reviewers have been asking
for tests that weren't there before when code gets refactored.

Over the next year or so, I hope to update *every* autodiff test in
Stan using the testing framework to improve test coverage and reduce
test code size and complexity.  I'll translate existing tests as I go, 
so nothing should get lost that's not done in the framework now.

> On Sep 4, 2019, at 4:34 AM, Rok Češnovar <notifications@github.com> wrote:
> 
> The tests for rev/mat/funinverse are already in place: https://github.com/stan-dev/math/blob/develop/test/unit/math/rev/mat/fun/inverse_test.cpp
> 
> They dont use your new testing framework yet.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

SteveBronder commented:
> I'm not reviewing, but I wouldn't accept this as is.  The tests are missing just about everything:
Tests for

 - [ ] behavior on NaN and infinite components
 - [ ] behavior on 0 x 0 and 1 x 1 matrices
 - [ ] exception behavior on non-square matrices
 - [ ] any other exception behavior on inputs (don't know if there is anything else)
 - [ ] behavior of derivatives (as is, being tested in the x[0, 1] slot for a 2x2 matrix
  and in testing gradients are all 1 in sum(inverse(inverse(x)) (because it's
  equivalent to sum(x)---would've saved me some time if that was documented)

> I know they were missing before the refactor, but reviewers have been asking
for tests that weren't there before when code gets refactored.

Thanks Bob!  Yeah my bad, I expected the "level" here for the contributor was the tests we previously imposed on the math library. But it's pretty fair to ask for better tests along with new functions.

Ivan would you mind updating the inverse reverse mode tests? It's mostly just using the standard error checks we have and catching the invalid argument, domain error, etc. There's a couple other tests with the finite difference method check you can use as a base.

> Over the next year or so, I hope to update *every* autodiff test in
Stan using the testing framework to improve test coverage and reduce
test code size and complexity.  I'll translate existing tests as I go,
so nothing should get lost that's not done in the framework now.

+1!

IvanYashchuk commented:
Sure, I will add those.

bbbales2 commented:
Here's the link for the new autodiff checks: https://github.com/stan-dev/math/wiki/Automatic-Differentiation-Testing-Framework#example-matrix-inverse-unary-function

Definitely do that and post if you have problems.

bob-carpenter commented:
> On Sep 4, 2019, at 3:08 PM, Steve Bronder <notifications@github.com> wrote:
> 
> ...

Bob:
> I know they were missing before the refactor, but reviewers have been asking
> for tests that weren't there before when code gets refactored.

Steve:
> Thanks Bob! Yeah my bad, I expected the "level" here for the contributor was the tests we previously imposed on the math library. But it's pretty fair to ask for better tests along with new functions.

This isn't my call, by the way, but I agree with Daniel here.  The worry is
that changes might break something that doesn't have good tests.  The feeling
is that inverse() is probably OK now because we haven't had user complaints about it.

I'm happy to write tests for this using the new framework just to illustrate how
they work in the higher-order case.  I just might not have time in the next week
as I'm technically on vacation. :-)

bob-carpenter commented:
I can update the tests for this and resubmit.  It'd be nice to get in a proper autodiff version of inverse as we do use it in some places.

IvanYashchuk commented:
I've put it off too long. I need to come back to this and learn the testing framework finally.

bob-carpenter commented:
If you want to do it, the test framework is in test/unit/math/test_ad.hpp and test_ad_test.cpp has a unit test involving matrix inverse which should be easy to copy.  

I can do it for you if you don't have time.  Let me know.  And thanks for submitting this. Sorry it's taken so long to circle back.  We have a new release coming up so I've been looking at what needs to get pushed through.

IvanYashchuk commented:
Added tests for
- [x]  input with NaN and infinite components
- [x]  input of size 0 x 0 and 1 x 1
- [x]  exception behavior on non-square matrices
- [x]  singular matrix input. NumPy (and LAPACK) would raise an error for this input, but Eigen does not do that.

When 0 x 0 input matrix is used Eigen internal assertion is raised, I added a check for that with nicer error message.

IvanYashchuk commented:
Windows Headers & Unit failed.
```
Recording test results
Remote call on EC2 (Windows) - Build, Docker, WSL (i-0f0f6a5efee3cc26b) failed
```
What's the reason for that?

wds15 commented:
We are seeing that on probably all PRs right now. Unstable Windows... @serban-nicusor-toptal could you take a look?

serban-nicusor-toptal commented:
@wds15 Looking into it.

bob-carpenter commented:
I moved the tests over to the new test framework and pushed changes. I added tests for 1 x 1, 2 x 2, 3 x 3, and 4 x 4 matrices at all levels of autodiff based on existing tests. It's kicking off tests.  I hope you realized I could push to your branch once you created a PR!  

Thanks again for submitting this.  I'll merge when tests pass.  

I realized while doing this that none of our linear algebra algorithms are set up to match tests for NaN in primitive and autodiff implementations.   A unifying decision needs to be made on how to approach edge values going forward and then prim, rev, fwd, mix implementations updated to match that decision.  When that happens we can test with the test framework.  As is, there were no tests at all for NaN behavior, so I felt OK just updating tests as is.

bob-carpenter commented:
I somehow missed all the above before blowing away tests and replacing with framework.  I didn't see the tests for sizes.  These need to be merged again.  Feel free to just back out what I did.  I'm not good enough at Git to do that myself.  Then I can approve and more carefully add conversions to the test framework later.  Sorry about mangling this PR.

bob-carpenter commented:
I think what I'm doing is leaving windows open in my browser, acting on those, then updating and only later seeing a bunch of new stuff.  I queued this one up to contribute to yesterday and changes came in between then and now and I didn't see them in my 200+ message/day email bomb from GitHub.

bob-carpenter commented:
Rather than risk messing up GitHub if someone else is in progress, I went back and looked at the commits from @IvanYashchuk and think the following should cover everything.  It also indicates some places where our API falls short, but that was all existing problems.  This PR makes it better, but not perfect.  And I document where things need to be fixed in the future.

```
#include <test/unit/math/test_ad.hpp>
#include <limits>

TEST(mathMixMatFun, inverse) {
  auto f = [](const auto& x) {
    return stan::math::inverse(x);
  };

  Eigen::Matrix<stan::math::var, -1, -1> t(0, 0);
  EXPECT_THROW(stan::math::inverse(t), std::invalid_argument);

  // fails with assertion from Eigen
  // stan::test::expect_ad(f, t);

  // fails with assertion from Eigen
  // Eigen::MatrixXd t2(0, 0);
  // EXPECT_THROW(stan::math::inverse(t2), std::invalid_argument);

  Eigen::MatrixXd u(1, 1);
  u << 2;
  stan::test::expect_ad(f, u);

  Eigen::MatrixXd v(2, 2);
  v << 2, 3, 5, 7;
  stan::test::expect_ad(f, v);
  v << 1.9, 0.3, 0.3, 1.7;
  stan::test::expect_ad(f, v);

  // issues around zero require looser tolerances for hessians
  stan::test::ad_tolerances tols;
  tols.hessian_hessian_ = 0.01;
  tols.hessian_fvar_hessian_ = 0.01;

  Eigen::MatrixXd w(3, 3);
  w << 2, 3, 5, 7, 11, 13, 17, 19, 23;
  stan::test::expect_ad(tols, f, w);

  // even lower tolerance, again for cases around zero
  stan::test::ad_tolerances tols2;
  tols2.hessian_hessian_ = 3.0;
  tols2.hessian_fvar_hessian_ = 3.0;

  Eigen::MatrixXd x(4, 4);
  x << 2, 3, 4, 5, 9, -1, 2, 2, 4, 3, 7, -1, 0, 1, 19, 112;
  stan::test::expect_ad(tols2, f, x);

  Eigen::MatrixXd y(3, 2);
  y << 1, 2, 3, 4, 5, 6;
  EXPECT_THROW(stan::math::inverse(y), std::invalid_argument);
  stan::test::expect_ad(f, y);

  Eigen::MatrixXd z(2, 2);
  z << 1, 2, 5, std::numeric_limits<double>::quiet_NaN();
  EXPECT_NO_THROW(stan::math::inverse(z));

  // autodiff throws, so following fails (throw behavior must match to pass)
  // stan::test::expect_ad(f, z);

  Eigen::MatrixXd a(2, 2);
  a << 1.9, 0.3, 0.3, std::numeric_limits<double>::infinity();
  stan::test::expect_ad(f, a);
}
```

IvanYashchuk commented:
> I hope you realized I could push to your branch once you created a PR!

Yes, I know that, it's possible as long as the person who submitted PR allows it.

I've updated the tests with the above suggestions. Thank you @bob-carpenter!
